### PR TITLE
Batch Verify Blocks

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

- [x] Adds ability to batch verify blocks from a peer.
- [x] Add feature config to utilize new feature.
- [x] Refactor some block processing methods to allow easier batch verification.
- [x] Add Tests

**Which issues(s) does this PR fix?**

**Other notes for review**

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	"encoding/hex"
 	"fmt"
 
+	"github.com/prysmaticlabs/prysm/shared/bls"
+
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:9]: spaces

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -245,7 +302,57 @@ func (s *Service) onBlockInitialSyncStateTransition(ctx context.Context, signed
 		}
 	}
 
-	// Epoch boundary bookkeeping such as logging epoch summaries.
+	return s.handleEpochBoundary(postState)
+}
+
+// handles a block after the block's batch has been verified, where we can save blocks
+// their state summaries and split them off to relative hot/cold storage.
+func (s *Service) handleBlockInBatch(ctx context.Context, signed *ethpb.SignedBeaconBlock,
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:310]:
```suggestion
func (s *Service) handleBlockAfterBatchVerify(ctx context.Context, signed *ethpb.SignedBeaconBlock,
```

`handleBlockInBatch` is not accurate since we only pass in a block as argument

File: beacon-chain/blockchain/process_block.go [hunk 3]
```diff
@@ -84,7 +86,7 @@ func (s *Service) onBlock(ctx context.Context, signed *ethpb.SignedBeaconBlock,
 		return nil, errors.Wrapf(err, "could not save block from slot %d", b.Slot)
 	}
 
-	if err := s.insertBlockToForkChoiceStore(ctx, b, blockRoot, postState); err != nil {
+	if err := s.insertBlockAndAttestationsToForkChoiceStore(ctx, b, blockRoot, postState); err != nil {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:89]: Why did we change this to `insertBlockAndAttestationsToForkChoiceStore`? We dont need to account for attestations here (additional overhead)

File: beacon-chain/blockchain/process_block.go [hunk 4]
```diff
@@ -202,17 +204,72 @@ func (s *Service) onBlockInitialSyncStateTransition(ctx context.Context, signed
 	if err != nil {
 		return errors.Wrap(err, "could not execute state transition")
 	}
+	return s.handlePostStateInSync(ctx, signed, blockRoot, postState)
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:207]: Did we need this before?

File: beacon-chain/blockchain/process_block.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/blockchain/process_block.go:89]: Just to clarify is there any reason we insert blocks/attestation into fork-choice in initial sync ? The only reason I split it this way was since, I assumed we did need to do it and account for both when syncing as that is the current behavior now.

File: beacon-chain/blockchain/process_block.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/blockchain/process_block.go:207]: This was to make it easier to break up these large functions , so that I could reuse portions of them when batch verifying them. Also these are currently very long methods which do a lot of things, which makes it hard to follow what is going on in them.

File: beacon-chain/blockchain/process_block.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/blockchain/process_block.go]: alright done

File: beacon-chain/blockchain/process_block.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:88]: Block is a must, attestation is only optional. But it was so cheap, we ended up doing both anyway to avoid disparity. I don't feel strongly about either, but would prefer clarity (lesser lines and methods). You can decide!
